docs/design/email_notification.md

```diff
@@ -0,0 +1,47 @@
+Email
+---
+
+#### Considerations
+
+- Be able to send frontpage email notifications on a daily or weekly basis
+  - Frequency is user-configurable
+  - If we don't have any new content since the last send, defer until next interval
+
+
+- Be able to notify users on specific events, controlled by a subscription to the entity
+  - Examples:
+    - Comment replies to posts
+    - Comment replies to other comments
+  - Capability to notify at the following frequencies:
+    - Near term we want to support:
+      - Never
+      - Immediate
+    - Long term we may want to support:
+      - Daily (digest/aggregation of new events)
+      - Weekly (digest/aggregation of new events)
+
+
+#### Proposal
+
+For frontpage digests the workflow would be:
+
+ - Daily and weekly tasks to:
+  - Walk `NotificationSettings` matching the corresponding `notification_type`
+  - For each of these settings, get frontpage data for that user
+  - Filter the list of frontpage posts to those that were made since the last frontpage email we sent this user
+    - A record needs to be stored at send time of when that happened
+  - If we have at least 1 post to send, render and send the emails
+  - Long a record of the sent email
+
+For event-specific notifications:
+
+ - When an event happens, trigger a django signal (so we can tie multiple handlers into it down the road)
+ - On that event, fire a celery task to check subscriptions
+ - Walk all subscriptions to that object and for each one create a `Notification` record
+   - If their `NotificationSettings` record for that type has `frequency_type` of `never`, then we don't create this record. This has the effect of muting subscriptions temporarily with necessitating the removal of the subscriptions or effecting a massive catch-up of notifications if they turn it back on
```

Comment:
Just a thought, but we may want to create a `Notification` even if the user has `frequency_type == 'never'` so that we don't create difficulties for ourselves later-on if we want to start showing these things in the UI. `Notification` could have a `send_mail` or `muted` boolean that would skip sending mail, but keep the notification itself around so that we could build a REST api to expose a list of notifications, their 'read' status, etc, if we want to go that route later.

Alternatively, we could go with the simpler thing now and add this complication later, it doesn't seem like something that would be terrible hard to add.

Reply:
That would definitely make sense for in-app notifications, but I'm trying to keep this design doc specific to the nuances of email notifications since in-app is out of scope for this initial implementation (although I'm trying hard not to design into a corner in that regard). We could always specify this to be an `EmailNotification` and then add a `InAppNotification` later and handle that differently.

That said, I think we ***really*** need some UI design around in-app notifications before proceeding with backend design though, because how I would design for in-app notifications hinges a lot on what the user experience is like, otherwise there's risk of designing something that's clunky or needs to be rewritten for our UX use cases.

Reply:
Cool, I think that makes sense, just wanted to make sure we weren't going to make a decision now that boxes us in later